### ISSUE TITLE
Feature Request: Specify Availability Zones when initializing environment

### ISSUE THREAD

**Issue by otariidae** [aws/copilot-cli#1298]
First off huge thanks for Copilot.

When I ran `copilot env init` with Load Balanced Web Service, copilot failed with the following error:
```
The following Availability Zones ap-northeast-1a cannot be associated with a load balancer
```
CloudFormation generated by copilot logged the same error.
CLI Version: 0.3.0

I figured out that `ap-northeast-1a` is mapped to `apne1-az3` and seems to have no support for Application Load Balancer.
It would be neccessary to have the options to specify Availability Zones to avoid this kind of error.

Workaround: Create a VPC and subnets in your own and select `No, I'd like to import existing resources (VPC, subnets)`.

**Comment by efekarakus** [comment]
ooh this is very interesting, thanks for also writing up a workaround! We'll update the issue as we make progress on the feature request :)

**Comment by Envek** [comment]
Faced this issue too:
```
The following Availability Zones ap-northeast-1b cannot be associated with a load balancer
```

Can it be related to access rights? Because I successfully deployed that same application in my personal account in the same `ap-northeast-1` region two-three weeks before, but it refuses to deploy into organization account (however, IAM role said to have admin access).

<details>
<summary>Copilot logs</summary>

1. `app init `

```
$ copilot app init panel-discussions
✔ Created the infrastructure to manage services under application panel-discussions.

✔ The directory copilot will hold service manifests for application panel-discussions.

Recommended follow-up actions:
- Run `copilot init` to add a new service to your application.
```

2. `copilot env init --name test`

```
$ copilot env init --name test
Which credentials would you like to use to create test? [profile ****-stg]
Would you like to use the default configuration for a new environment?
    - A new VPC with 2 AZs, 2 public subnets and 2 private subnets
    - A new ECS Cluster
    - New IAM Roles to manage services in your environment
 Yes, use default.
✘ Failed to create the infrastructure for the test environment.
- Virtual private cloud on 2 availability zones to hold your services     [Complete]
- Virtual private cloud on 2 availability zones to hold your services                [Complete]
  - Internet gateway to connect the network to the internet                          [Complete]
  - Public subnets for internet facing services                                      [Complete]
  - Private subnets for services that can't be reached from the internet             [Complete]
  - Routing tables for services to talk with each other                              [Failed]
  Resource creation cancelled                                                                 
- ECS Cluster to hold your services                                                  [Complete]
- Application load balancer to distribute traffic                                    [Failed]
  The following Availability Zones ap-northeast-1b cannot be associated with a load balancer  
✘ wait until stack panel-discussions-test create is complete: ResourceNotReady: failed waiting for successful resource state
```
</details>
<details>
<summary>Error in Cloudformation</summary>

```
The following Availability Zones ap-northeast-1b cannot be associated with a load balancer. Please try a different Availability Zone. (Service: AmazonElasticLoadBalancing; Status Code: 400; Error Code: ValidationError; Request ID: 0b1c17cc-770a-4108-b673-7d095b67b388; Proxy: null)
```

![image](https://user-images.githubusercontent.com/264400/96573330-a6763c80-12d6-11eb-8523-db3cfe3a858b.png)
</details>

</details>

**Comment by Envek** [comment]
What is interesting, my previous attempt to create environment in my personal account successfully created load balancer in availability zones `ap-northeast-1c` and `ap-northeast-1a`.
<details>
<summary>See screenshot</summary>

![image](https://user-images.githubusercontent.com/264400/96584018-46879200-12e6-11eb-9115-b742d5c10eb1.png)
</details>

But as far as I can understand from [Copilot manifest](https://github.com/aws/copilot-cli/blob/aeb2e12dc98fa88a04c6ca704ea21df8d83afcf5/templates/environment/partials/vpc-resources.yml) and [GetAZs function docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/intrinsic-function-reference-getavailabilityzones.html), it shouldn't happen because:

> ` Fn::GetAZs` returns an array that lists Availability Zones for a specified region in alaphabetical order

But Copilot tooks only first two AZs from it:
https://github.com/aws/copilot-cli/blob/aeb2e12dc98fa88a04c6ca704ea21df8d83afcf5/templates/environment/partials/vpc-resources.yml#L40-L46
https://github.com/aws/copilot-cli/blob/87308f4a50fe22dd3e8c85b9ea567bdca51053de/templates/environment/versions/cf-v1.0.0.yml#L124

Which means that `Fn::GetAZs` didn't return `ap-northeast-1b` at the time of environment creation in my personal account (October 6, 2020)

And this is because `ap-northeast-1b` isn't enabled for my personal account but enabled for organization account:
<details>
<summary>My personal account AZs: ap-northeast-1a (apne1-az4), ap-northeast-1c (apne1-az1), ap-northeast-1d (apne1-az2)</summary>

![image](https://user-images.githubusercontent.com/264400/96586590-e1ce3680-12e9-11eb-9be8-27ef0e31dbbd.png)
</details>
<details>
<summary>Organization account AZs: ap-northeast-1a (apne1-az4), ap-northeast-1b (apne1-az3), ap-northeast-1c (apne1-az1), ap-northeast-1d (apne1-az2)</summary>

![image](https://user-images.githubusercontent.com/264400/96586740-117d3e80-12ea-11eb-9e6b-e10efd275cde.png)

</details>

**Comment by otariidae** [comment]
FYI:

`ap-northeast-1b` is an Availability Zone. Availability Zones could differ with AWS accounts.

> For example, the Availability Zone us-east-1a for your AWS account might not be the same location as us-east-1a for another AWS account.
> https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/using-regions-availability-zones.html#concepts-availability-zones

`apne1-az3` is an AZ ID. AZ IDs are unique in every AWS accounts.

> For example, use1-az1 is an AZ ID for the us-east-1 Region and it has the same location in every AWS account.
> https://docs.aws.amazon.com/ram/latest/userguide/working-with-az-ids.html

To see AZ IDs for your AZs run `aws ec2 describe-availability-zones`.

`apne1-az3` does not support Application Load Balancer.

Newer created AWS accounts may not have the AZ for `apne1-az3`.

**Comment by Envek** [comment]
Thank you for clarification! AZ names are equal in my personal and organizational accounts.

Is there a way to somehow disable certain AZ in account? I can't see such a possibility…

**Comment by otariidae** [comment]
No way to disable certain AZ.
Copilot automatically select AZs by default even if they contain `apne1-az3`.

It boils down to [my first comment](https://github.com/aws/copilot-cli/issues/1298#issue-681729120) if you use Copilot v0.3.0.

> Workaround: Create a VPC and subnets in your own and select `No, I'd like to import existing resources (VPC, subnets)`.

Copilot v0.4.0 can import existing subnets:

```
$ copilot env init -h
(...)
Import Existing Resources Flags
      --import-private-subnets strings   Optional. Use existing private subnet IDs.
      --import-public-subnets strings    Optional. Use existing public subnet IDs.
(...)
```
https://aws.github.io/copilot-cli/docs/commands/env-init/

**Comment by Envek** [comment]
Yes, I even copy-pasted VPC and subnets creation logic from Copilot and created the following CloudFormation template to automate this:
<details>
<summary>copilot/custom-vpc-subnets.yml</summary>

```yml
# Taken from https://github.com/aws/copilot-cli/blob/87308f4a50fe22dd3e8c85b9ea567bdca51053de/templates/environment/partials/vpc-resources.yml
# (see https://github.com/aws/copilot-cli/blob/87308f4a50fe22dd3e8c85b9ea567bdca51053de/templates/environment/versions/cf-v1.0.0.yml#L38-L40)
# to work around https://github.com/aws/copilot-cli/issues/1298
#
# Run with:
# app=panel-discussions
# env=test
# aws cloudformation create-stack --stack-name $app-$env-vpc-subnets --template-body file://copilot/custom-vpc-subnets.yml --parameters ParameterKey=App,ParameterValue=$app ParameterKey=Env,ParameterValue=$env --tags Key=copilot-application,Value=$app Key=copilot-environment,Value=$env

Parameters:
  App:
    Type: String
    Description: Your application's name.
  Env:
    Type: String
    Description: The copilot environment name.
  AZs:
    Type: CommaDelimitedList
    Default: ap-northeast-1a, ap-northeast-1c
    # AllowedValues: ["ap-northeast-1a", "ap-northeast-1c", "ap-northeast-1d"]
    # Check output of `aws ec2 describe-availability-zones` command and exclude name of AZ with ID `apne1-az3` from it (it was `ap-northeast-1b` in my case).
  CIDR:
    Type: String
    Default: 10.0.0.0/16
  PublicCIDRs:
    Type: CommaDelimitedList
    Default: 10.0.0.0/24, 10.0.1.0/24
  PrivateCIDRs:
    Type: CommaDelimitedList
    Default: 10.0.2.0/24, 10.0.3.0/24

Resources:
  VPC:
    Type: AWS::EC2::VPC
    Properties:
      CidrBlock: !Ref CIDR
      EnableDnsHostnames: true
      EnableDnsSupport: true
      InstanceTenancy: default
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}'

  PublicRouteTable:
    Type: AWS::EC2::RouteTable
    Properties:
      VpcId: !Ref VPC
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}}'

  DefaultPublicRoute:
    Type: AWS::EC2::Route
    DependsOn: InternetGatewayAttachment
    Properties:
      RouteTableId: !Ref PublicRouteTable
      DestinationCidrBlock: 0.0.0.0/0
      GatewayId: !Ref InternetGateway

  InternetGateway:
    Type: AWS::EC2::InternetGateway
    Properties:
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}}'

  InternetGatewayAttachment:
    Type: AWS::EC2::VPCGatewayAttachment
    Properties:
      InternetGatewayId: !Ref InternetGateway
      VpcId: !Ref VPC

  PublicSubnet1:
    Type: AWS::EC2::Subnet
    Properties:
      CidrBlock: !Select [ 0, !Ref PublicCIDRs ]
      VpcId: !Ref VPC
      AvailabilityZone: !Select [ 0, !Ref AZs ]
      MapPublicIpOnLaunch: true
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}}-pub0'

  PublicSubnet2:
    Type: AWS::EC2::Subnet
    Properties:
      CidrBlock: !Select [ 1, !Ref PublicCIDRs ]
      VpcId: !Ref VPC
      AvailabilityZone: !Select [ 1, !Ref AZs ]
      MapPublicIpOnLaunch: true
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}}-pub1'

  PrivateSubnet1:
    Type: AWS::EC2::Subnet
    Properties:
      CidrBlock: !Select [ 0, !Ref PrivateCIDRs ]
      VpcId: !Ref VPC
      AvailabilityZone: !Select [ 0, !Ref AZs ]
      MapPublicIpOnLaunch: false
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}}-priv0'

  PrivateSubnet2:
    Type: AWS::EC2::Subnet
    Properties:
      CidrBlock: !Select [ 1, !Ref PrivateCIDRs ]
      VpcId: !Ref VPC
      AvailabilityZone: !Select [ 1, !Ref AZs ]
      MapPublicIpOnLaunch: false
      Tags:
        - Key: Name
          Value: !Sub 'copilot-${App}-${Env}}-priv1'

  PublicSubnet1RouteTableAssociation:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId: !Ref PublicRouteTable
      SubnetId: !Ref PublicSubnet1

  PublicSubnet2RouteTableAssociation:
    Type: AWS::EC2::SubnetRouteTableAssociation
    Properties:
      RouteTableId: !Ref PublicRouteTable
      SubnetId: !Ref PublicSubnet2

Outputs:
  VPC:
    Value: !Ref VPC
  PublicSubnet1:
    Value: !Ref PublicSubnet1
  PublicSubnet2:
    Value: !Ref PublicSubnet2
  PrivateSubnet1:
    Value: !Ref PrivateSubnet1
  PrivateSubnet2:
    Value: !Ref PrivateSubnet2
```
</details>

And invoke it with following command:

```sh
app=panel-discussions
env=test
aws cloudformation create-stack --stack-name $app-$env-vpc-subnets \
  --template-body file://copilot/custom-vpc-subnets.yml \
  --parameters ParameterKey=App,ParameterValue=$app ParameterKey=Env,ParameterValue=$env \
  --tags Key=copilot-application,Value=$app Key=copilot-environment,Value=$env
```

And then create environment itself:

```sh
copilot env init --name test
```

Choose `No, I'd like to import existing resources (VPC, subnets).` and choose VPC created by CloudFormation stack on the previous command (or provide them via `--import-private-subnets` and `--import-private-subnets` options).

**Comment by toricls** [comment]
Pasting a tweet who might be facing the exact same issue here 📝 - https://twitter.com/elliottkember/status/1361159350480297985

**Comment by huanjani** [comment]
This is now released 🎉 in v1.14.0 https://github.com/aws/copilot-cli/releases/tag/v1.14.0 !
